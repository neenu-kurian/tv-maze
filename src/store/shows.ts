import {
  fetchShowsByCategory,
  fetchShowData,
  fetchCastData,
  searchShows,
} from '@/services/index.ts';
import type {
  CategorizedShows,
  ShowsState,
  CustomError,
  Show,
  CastMember,
  CastError,
} from '@/types/index';
import { defineStore } from 'pinia';

export const useShowsStore = defineStore('shows', {
  state: (): ShowsState => ({
    showsByCategory: {},
    showDetails: {},
    loading: false,
    searchQuery: '',
    sortBy: 'rating-desc',
    searchResult: [],
    searchError: false,
  }),
  getters: {
    sortedShows(): CategorizedShows | CustomError {
      if ('error' in this.showsByCategory || !this.showsByCategory) {
        return this.showsByCategory;
      }
      const showsToSort = { ...this.showsByCategory };
      Object.entries(showsToSort).forEach(([genre, shows]) => {
        showsToSort[genre] = [...shows].sort((a, b) => {
          const ratingA = a.rating?.average || 0;
          const ratingB = b.rating?.average || 0;
          return this.sortBy === 'rating-asc' ? ratingA - ratingB : ratingB - ratingA;
        });
      });
      return showsToSort;
    },
    shows() {
      return this.sortedShows;
    },
  },
  actions: {
    async searchShows(query: string) {
      this.loading = true;
      this.searchQuery = query;
      this.searchResult = await searchShows(query);
      this.loading = false;
    },
    async getShowsByCategory() {
      this.loading = true;
      this.showsByCategory = await fetchShowsByCategory();
      this.loading = false;
    },
    setSearchText(text: string) {
      this.searchQuery = text;
    },
    setShowDetails(show: Show | CustomError) {
      if (show && 'id' in show && show.id !== undefined) {
        this.showDetails[show.id] = show;
      }
    },
    setCastDetails(cast: CastMember[] | CastError, id: number) {
      if (this.showDetails[id] && !('error' in this.showDetails[id])) {
        this.showDetails[id].cast = cast;
      }
    },
    async fetchShowDetails(id: number): Promise<Show | CustomError> {
      this.loading = true;
      const show = await fetchShowData(id);
      this.loading = false;
      this.setShowDetails(show);
      return show;
    },

    async fetchCastDetails(id: number): Promise<CastMember[] | CastError> {
      this.loading = true;
      const cast = await fetchCastData(id);
      this.loading = false;
      this.setCastDetails(cast, id);
      return cast;
    },
  },
});
