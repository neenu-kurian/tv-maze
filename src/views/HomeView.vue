<template>
  <div>
    <header class="sticky top-0 bg-white py-6 -mt-8 mb-8 flex flex-col gap-4 shadow-md z-1">
      <h1 class="text-3xl font-bold text-gray-900 md:text-left md:ml-5 text-center">TV Maze</h1>
      <div class="md:flex items-center gap-16 w-full px-5 block">
        <SearchInput
          v-model="searchQuery"
          @enter="handleSearch"
          @update:modelValue="handleSearch"
        />
        <SortDropdown v-model="sortBy" :options="sortOptions" class="mt-5 md:mt-0" />
      </div>
    </header>

    <div class="pl-10 mt-15">
      <template v-if="isSearching">
        <SearchResults :search-query="searchQuery" :shows="searchResult" :loading="searchLoading" />
      </template>

      <template v-else>
        <AppLoader v-if="loading" :message="'Loading shows...'" />
        <template v-else>
          <template v-if="error">
            <div>
              <p>{{ error }}</p>
            </div>
          </template>
          <template v-else>
            <template v-if="Object.keys(sortedShows).length === 0">
              <div>
                <p>No shows found</p>
              </div>
            </template>
            <template v-else>
              <template v-for="(showsByGenre, genre) in sortedShows" :key="genre">
                <section>
                  <div class="flex items-center gap-2 mb-4 align-middle">
                    <h2 class="text-xl font-bold text-gray-900 m-0">{{ genre }}</h2>
                    <ArrowRightIcon class="w-5 h-5 text-gray-600 relative top-1/2" />
                  </div>
                  <div role="list" class="flex gap-5 overflow-x-auto pb-4 align-stretch max-w-[94%]">
                    <ShowCard v-for="(show, index) in showsByGenre" :key="index" :show="show" />
                  </div>
                </section>
              </template>
            </template>
          </template>
        </template>
      </template>
    </div>
  </div>
</template>

<script setup lang="ts">
import { onMounted, ref, computed, onUnmounted } from 'vue';
import ShowCard from '@/components/ShowCard.vue';
import SearchResults from '@/components/SearchResults.vue';
import { useShowsStore } from '@/store/shows.ts';
import { storeToRefs } from 'pinia';
import SearchInput from '@/components/SearchInput.vue';
import AppLoader from '@/components/AppLoader.vue';
import SortDropdown from '@/components/SortDropdown.vue';
import { sortOptions } from '@/constants.ts';
import { ArrowRightIcon } from '@heroicons/vue/24/outline';

const showsStore = useShowsStore();
const { loading, sortedShows, sortBy, showsByCategory, searchQuery, searchResult } =
  storeToRefs(showsStore);
const error = ref<string | null>(null);
const searchLoading = ref(false);
const isSearching = computed(() => searchQuery.value.trim() !== '');

let searchTimeout: number | null = null;

const handleSearch = async () => {
  if (searchTimeout) {
    clearTimeout(searchTimeout);
  }
  searchLoading.value = true;
  searchTimeout = setTimeout(async () => {
    if (searchQuery.value.trim() === '') {
      clearSearch();
      searchLoading.value = false;
      return;
    }
    await showsStore.searchShows(searchQuery.value).then(() => {
      searchLoading.value = false;
    });
  }, 1000);
};

const clearSearch = () => {
  searchQuery.value = '';
  showsStore.setSearchText('');
};

onMounted(async () => {
  loading.value = true;
  if ('error' in showsByCategory.value) {
    error.value = 'Failed to load shows. Please try again later.';
    return;
  }
  if (Object.keys(showsByCategory.value).length === 0) {
    await showsStore.getShowsByCategory();
  }
  loading.value = false;
});

onUnmounted(() => {
  if (searchTimeout) {
    clearTimeout(searchTimeout);
  }
});
</script>
