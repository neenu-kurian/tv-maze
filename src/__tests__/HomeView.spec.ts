import { render } from '@testing-library/vue';
import HomeView from '@/views/HomeView.vue';
import { describe, it, expect, beforeEach } from 'vitest';
import { createPinia, setActivePinia } from 'pinia';
import { useShowsStore } from '@/store/shows';
import { showsByCategory } from './testdata.ts';

describe('HomeView.vue', () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });
  it('matches snapshot', () => {
    const showsStore = useShowsStore();
    showsStore.$patch({
      showsByCategory,
    });
    const { container } = render(HomeView, {
      global: {
        stubs: {
          RouterLink: true,
        },

      },
    });
    expect(container).toMatchSnapshot();
  });
});
