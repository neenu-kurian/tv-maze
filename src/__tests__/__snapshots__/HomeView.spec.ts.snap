// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeView.vue > matches snapshot 1`] = `
<div>
  <div>
    <header
      class="sticky top-0 bg-white py-6 -mt-8 mb-8 flex flex-col gap-4 shadow-md z-1"
    >
      <h1
        class="text-3xl font-bold text-gray-900 md:text-left md:ml-5 text-center"
      >
        TV Maze
      </h1>
      <div
        class="md:flex items-center gap-16 w-full px-5 block"
      >
        <div
          class="flex align-center gap-6 rounded-md border-transparent px-4 py-2 md:w-[60%] max-w-full bg-gray-100 width-full"
          role="search"
        >
          
          <svg
            aria-hidden="true"
            class="w-8 h-8 flex-none basis-5"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          
          <input
            class="border-none outline-none bg-transparent w-full text-base text-gray-900 p-0 leading-none"
            name="search-input"
            placeholder="Search for TV shows..."
            type="text"
            value=""
          />
        </div>
        <div
          class="relative inline-flex items-center bg-gray-100 rounded-lg px-3 md:min-w-[200px] float-right mt-5 md:mt-0"
        >
          <select
            class="appearance-none bg-transparent border-none py-2.5 text-sm text-gray-900 w-full cursor-pointer font-medium"
            name="select-sort"
            value="rating-desc"
          >
            
            <option
              value="rating-desc"
            >
              Rating: High to Low
            </option>
            <option
              value="rating-asc"
            >
              Rating: Low to High
            </option>
            
          </select>
          <svg
            aria-hidden="true"
            class="w-4 h-4 text-gray-500 ml-2 shrink-0 pointer-events-none"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m19.5 8.25-7.5 7.5-7.5-7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </header>
    <div
      class="pl-10 mt-15"
    >
      
      
      
      
      <section>
        <div
          class="flex items-center gap-2 mb-4 align-middle"
        >
          <h2
            class="text-xl font-bold text-gray-900 m-0"
          >
            Music
          </h2>
          <svg
            aria-hidden="true"
            class="w-5 h-5 text-gray-600 relative top-1/2"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.5 4.5 21 12m0 0-7.5 7.5M21 12H3"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
        <div
          class="flex gap-5 overflow-x-auto pb-4 align-stretch max-w-[94%]"
          role="list"
        >
          
          <div
            class="flex flex-col gap-2 mb-5"
          >
            <router-link-stub
              ariacurrentvalue="page"
              class="hover:-translate-y-1 transform transition"
              custom="false"
              replace="false"
              to="[object Object]"
              viewtransition="false"
            />
            <h3
              class="text-base font-medium mt-2 mb-0 text-gray-800 text-left px-2 overflow-hidden"
            >
              Nashville
            </h3>
          </div>
          <div
            class="flex flex-col gap-2 mb-5"
          >
            <router-link-stub
              ariacurrentvalue="page"
              class="hover:-translate-y-1 transform transition"
              custom="false"
              replace="false"
              to="[object Object]"
              viewtransition="false"
            />
            <h3
              class="text-base font-medium mt-2 mb-0 text-gray-800 text-left px-2 overflow-hidden"
            >
              Glee
            </h3>
          </div>
          
        </div>
      </section>
      
      
      
      
    </div>
  </div>
</div>
`;
